File: src/ActualLab.Rpc/Infrastructure/RpcCallStage.cs
namespace ActualLab.Rpc.Infrastructure;

public static class RpcCallStage
{
    private static readonly Lock Lock = new();
    private static Dictionary<int, string> _callStageNames = new() {
        { ResultReady, nameof(ResultReady) },
        { ResultReady | Unregistered, "*" + nameof(ResultReady) },
        { Invalidated, nameof(Invalidated) },
        { Invalidated | Unregistered, "*" + nameof(Invalidated) },
        { Unregistered, "*" + nameof(Unregistered) },
    };

    public const int ResultReady = 1;
    public const int Invalidated = 3;
    public const int Unregistered = 0x1_000;

    public static void Register(int value, string name)
    {
        lock (Lock) {
            var callStageNames = new Dictionary<int, string>(_callStageNames) {
                { value, name },
                { value | Unregistered, "*" + name },
            };
            _callStageNames = callStageNames;
        }
    }

    public static string GetName(int completedStage)
        => _callStageNames.GetValueOrDefault(completedStage) ?? "";
}

File: ts/packages/rpc/src/rpc-call-stage.ts
/**
 * Well-known RPC call completion stage constants, used by the
 * `$sys.Reconnect` protocol to tell the peer which calls have already
 * reached which stage and therefore don't need to be re-processed.
 *
 * Direct port of .NET `ActualLab.Rpc.Infrastructure.RpcCallStage` at
 * src/ActualLab.Rpc/Infrastructure/RpcCallStage.cs.
 */
export const RpcCallStage = {
    /** The call's result has been received from the peer. */
    ResultReady: 1,
    /** The compute call's result has been invalidated (compute-call specific). */
    Invalidated: 3,
    /**
     * Bit flag added on top of ResultReady/Invalidated when the outbound
     * call has been removed from the tracker (unregistered).
     */
    Unregistered: 0x1_000,
} as const;

export type RpcCallStage = number;
